### DISCUSSION TITLE
Plugin Library

### DISCUSSION THREAD

**Discussion by Drallas** [jhuckaby/Cronicle#652] [General]
Perhaps I missed it, but is there somewhere a Library with community plugins? Or are the built in ones the only available and it’s to me to create my own!?

**Comment by jhuckaby** [comment]
There's no community plugin library that I am aware of.

Custom Plugins are only for very specific and advanced requirements.  Almost everyone just uses the built-in [Shell Plugin](https://github.com/jhuckaby/Cronicle/blob/master/docs/Plugins.md#built-in-shell-plugin), which can run any commands or even exec full scripts in any scripting language.  It should do anything you need.

**Comment by Drallas** [comment]
ok, thanks, so there is no library with awesome plugins somewhere!

**Reply by mikeTWC1984** [reply]
@Drallas  Shell Plugin is a swiss army knife indeed. Keep in mind that it support some json messaging option (e.g. you can chain a job with some extra info, add some html to log etc)

@jhuckaby  some related thoughts on that:
 - Turns out that you can set up RPC channel between parent and child node processes (e.g. when spawning shell plugin). This way child and parent can exchange with messages (commands) in real time (not just at the end). I think that might be an great tool to build complex plugins.(like workflows). Implementation is also very simple just need to add 'ipc' to stdio option (stdio: [pipe, pipe, pipe, 'ipc'). It also give an option to gracefully stop a background node process on windows (via disconnect function)
 - ShellPlug is most popular plugin (close to 100% for many users). However it's creating extra node process, so say if you just run bash command like "sleep 10", it will still eat up 40MB of memory. Perhaps it would make sense to bake shell plugin into cronicle somehow.

**Reply by Drallas** [reply]
Any sources for inspiration, how others use it?

**Comment by jhuckaby** [comment]
> Turns out that you can set up RPC channel between parent and child node processes

Orchestra actually already supports RPC (it's just JSON over STDIO, but you can send/receive any time during a job).  One thing this allows you to do is send real-time job updates to the UI, so you can show your own custom HTML progress, stats, that all update in real time.  It's pretty awesome.

> However it's creating extra node process, so say if you just run bash command like "sleep 10", it will still eat up 40MB of memory. Perhaps it would make sense to bake shell plugin into cronicle somehow.

Yup, that's actually already handled as well.  In Orchestra, there is a "satellite" that gets installed on all your servers where you want jobs executed (no need to install the full web app anymore 🎉).  This satellite is a compiled static binary (via Vercel's [pkg](https://www.npmjs.com/package/pkg)), and has zero dependencies.  As a result of this, there are no more secondary scripts or processes that it spawns for the built-in functionality of Shell / Web Requests / etc.  It all happens inside that one process, reducing the memory footprint, and allowing for SUPER easy satellite installs on servers that don't even need to have Node.js nor Orchestra installed.

**Comment by josephdadams** [comment]
@Drallas  here are the plugins I have written, but as @jhuckaby said, they're pretty custom for us. But maybe will inspire you for yours. https://github.com/josephdadams/cronicleplugins

**Reply by Drallas** [reply]
Thanks for the inspiration, didn't know yet I could use Python for that too!

**Reply by josephdadams** [reply]
You can use anything executable by the system! I use shebangs at the beginning of my files and chmod +x them to be executable.
